Feature/pure ruby ssl implementation for root certificate issuer check

jspaleta opened the pull request:
## Pull Request Checklist
This is to address use of openssl command line tool in the check-ssl-anchor with an alternative implementation of the same logic using pure base ruby packages.

This is an outgrowth of the clean up done in PR #70  

Probably should merge PR #70 first.
 
#### General

- [X] Update Changelog

- [X] RuboCop passes

- [X] Existing tests pass

#### New Plugins

- [X] Tests

- [X] Add the plugin to the README


#### Purpose
Provide full ruby implementation of similar logic to `check-ssl-anchor.rb` without requireing openssl cmdline tool.

#### Known Compatibility Issues

jspaleta committed:
Add option to treat anchor as a regexp. Fix parsing of openssl client output to work with both openssl 1.0 and openssl 1.1 formatting

jspaleta committed:
updates to make travis and rubocop happy

jspaleta committed:
Add pure ruby implementation of check-ssl-root-issuer.rb as alternative to check-ssl-anchor.rb

jspaleta committed:
make rubocop happy

jspaleta committed:
add test for check-ssl-root-issuer

jspaleta committed:
update changelog and README with new plugin information

majormoses commented on bin/check-ssl-root-issuer.rb:
```diff
@@ -0,0 +1,128 @@
+  def validate_issuer(cert)
+    issuer = cert.issuer.to_s(config[:issuer_format])
+    if config[:regexp]
+      ra = Regexp.new(config[:issuer].to_s)
```
can we have a better var name? I'd prefer longer more descriptive vars over shaving a few chars.

majormoses commented on bin/check-ssl-root-issuer.rb:
```diff
@@ -0,0 +1,128 @@
+  end
+
+  def find_root_cert(uri)
+    pcert = root_cert = nil
```
same comment, I'd prefer if we could have var names that do not require you to trace code to understand their intent. Is this parent, previous, etc cert.

majormoses commented on bin/check-ssl-root-issuer.rb:
```diff
@@ -0,0 +1,128 @@
+    @failed_cert_reason = 'Unknown'
+    validate_opts
+    uri = URI.parse(config[:url])
+    critical 'Url must be https' if uri.scheme != 'https'
```
```suggestion
    critical "Url protocol must be https, you specified: #{uri}" if uri.scheme != 'https'
```

majormoses commented on test/check-ssl-hsts-preloadable_spec.rb:
```diff
@@ -19,11 +19,11 @@
-    expect(check).to receive(:warning).and_raise SystemExit
-    expect { check.run }.to raise_error SystemExit
-  end
+  #  it 'should pass check if the domain is preloadedable but has warnings' do
```
if you are going to comment these out lets include a explanation as to why

phumpal replied:
Can we leave these commented? I'm still not 100% sure what's required here.

jspaleta replied:
addressed in new commit

jspaleta committed:
changes from pr review

jspaleta committed:
remove files changed in PR #70, unrelated to this new feature

jspaleta commented:
I've addressed the review items with latest commits.

I've also updated this branch to remove the files changed in PR #70 associated with the failing check-ssl-anchor test.   

Pr #70 should be merged first to clean up the failing anchor test, and then I'll rebase this PR off master

jspaleta committed:
make rubocop happy

majormoses commented on bin/check-ssl-root-issuer.rb:
```diff
@@ -0,0 +1,127 @@
+         default: 'RFC2253',
+         required: false
+
+  def validate_opts
```
A couple thoughts:
- I think we should we still add validation to the options passed in themselves. if I pass in an invalid value I see no reason to run this code here
- I think we should add conditional logic, if I specify something other than those two options the way the code is written it will attempt to validate with `RFC2253` regardless of what is passed in
- should we add a qualifier here that this is specific to format or do you intend that we would keep adding more validation to this?


Maybe that should be something we pass in and can try something like this? (not sure if this is valid)
```ruby
def validate_format_opts(config[:format])
  OpenSSL::X509::Name::config[:format]
end


# I am pretty sure this will work but is dangerous and could lead easily to code injection
# that being said if we validate the options beforehand I think that is sufficient mitigation
# (which I would call out in a comment) and will make sense right before the rubocop
# disable as I am sure it will (rightly so) yell at us as its incredibly dangerous but powerful
# if wielded properly
def validate_format_opts(config[:format])
  eval "OpenSSL::X509::Name::#{config[:format]}"
end
```
if we cant specify it like that then we could always use if, elif, else or case statements to handle.

jspaleta replied:
Okay I've updated the logic to use mixin's cli option  value validation method.

majormoses reviewed (commented):
Overall looks good, needs a rebase.

jspaleta committed:
Merge branch 'master' into feature/pure_ruby_ssl

jspaleta committed:
Update logic for validating issuer name format options. Using mixin libraries internal validation for allowed values.

majormoses commented:
@jspaleta let me know if you need any help appeasing the cops, in the case of the eval please do indicate in a comment that the validation is done at user input so should be safe in this case.

jspaleta committed:
make rubocop happy.

jspaleta commented:
Rubocop fixed.
Eval rubocop override commented.

majormoses commented on CHANGELOG.md:
```diff
@@ -11,12 +11,14 @@ This CHANGELOG follows the format listed [here](https://github.com/sensu-plugins
 
 ### Fixed
 - ssl-anchor test now uses regexp
+>>>>>>> master
```
leftover merge

majormoses reviewed (approved):
LGTM

One minor fixup in the changelog.

jspaleta committed:
clean up changelog

majormoses reviewed (approved):
LGTM

majormoses commented:
@jspaleta do you want to merge this and prep a release or do you want me to? I can probably get to it this evening if you don't get to it sooner.
